Fix computation of response length in the general case

odelalleau opened the pull request:
This is a resubmit of #4 after rebasing. ~~Note that I dropped the changelog update (dc0a979c) since the release is still pending.~~

shengyangs commented:
Is there any particular reason why `\x11` is not supported here? If I modify the code here to include `\x11`, will it incur some problem? https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/models/nlp/gpt/megatron_gpt_ppo_actor.py#L69

odelalleau commented:
> Is there any particular reason why `\x11` is not supported here? If I modify the code here to include `\x11`, will it incur some problem? https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/models/nlp/gpt/megatron_gpt_ppo_actor.py#L69

Yeah it will break unless you grab this PR. Sorry I had completely forgotten about this PR, I'll need to fully test it to merge it in, but it should work for you.

odelalleau commented:
> it should work for you.

Just a heads-up that I actually ran into some issues when trying it out, so it may not work just yet. Unfortunately I won’t be able to look into it until Jan. 4th.

shengyangs commented:
No worries. I managed to find another hacky solution. Yet still tuning RLHF performances. I will talk with you after you are back. Enjoy the holiday.

odelalleau commented:
I fixed the issue and rebased. Will test it more thoroughly now. The problem was that
```python
self._end_idx = torch.where(started & is_end & (self._end_idx < 0), context_length, self._end_idx)
```
was missing the parentheses around `self._end_idx < 0`

odelalleau commented:
Un-drafting as I think I have run enough experiments with this code to be confident it works fine:
* With byte tokens (`\x11`)
* With `<extra_id_1>` tokens
* With `<extra_id_1>` as text

gshennvm commented on nemo_aligner/utils/text_generation_utils.py:
```diff
@@ -14,11 +14,60 @@
+        self, tokens: torch.Tensor, prev: torch.Tensor, eod_id: int, end_strings: List[str]
+    ) -> torch.Tensor:
+        is_end = super().end_of_generation_condition(tokens=tokens, prev=prev, eod_id=eod_id, end_strings=end_strings)
+        assert len(is_end) == len(tokens)
+        if len(tokens) != len(self._context_lengths):
+            raise RuntimeError(
+                "Batch size mismatch: the `context_lengths` tensor provided in the constructor has batch size "
+                f"{len(self._context_lengths)}, while the generated tokens have batch size {len(tokens)}"
+            )
```
this logic looks to guard against the batch size changing, in what cases can this occur?

odelalleau replied:
In case someone instantiate a `TrackLengthGPTModelTextGenerationStrategy` with an incorrect `context_lengths`. Definitely not supposed to happen with the current code but since this is a standalone class (and I wanted to make sure my code was correct) I preferred to add a check.

gshennvm commented on nemo_aligner/models/nlp/gpt/megatron_gpt_ppo_actor.py:
```diff
@@ -280,19 +262,36 @@ def infer(self, inference_batch):
         )
 
+        response_lengths = None
+        if parallel_state.is_pipeline_last_stage():
```
(just a thought): the contract we have with nemo generate is that the outputs it returns will be duplicated across all ranks. 

This breaks in contract -- I think we should move this broadcast PP logic into the strategy itself(since the strategy is already checking if PP is last stage when computing get_length). 

then in the model we can just perform the `assert (response_lengths <= self.cfg.encoder_seq_length).all()` sanity check

odelalleau replied:
Good idea, I agree it's nice that way. Done in 02f0a612.

odelalleau commented on nemo_aligner/models/nlp/gpt/megatron_gpt_ppo_actor.py:
```diff
@@ -280,19 +262,36 @@ def infer(self, inference_batch):
-        )
+
+        # Sanity check to validate response length.
+        if max_response_length != response_tokens.size(1):
```
Hmm not sure how, this check uses the response tokens which the strategy doesn't keep track of. It's really just a sanity check to be sure the response length we compute is consistent with the generation output (that is not fully controlled by the strategy).

gshennvm replied:
ah right OK, let's leave it for now

gshennvm reviewed (approved):
thanks for the change! this LGTM. This also cleans up our logic on hard coding `<extra_id_1>` which is very welcomed

cc @JRD971000

gshennvm reviewed (dismissed):
thanks for the change! this LGTM. This also cleans up our logic on hard coding `<extra_id_1>` which is very welcomed

cc @JRD971000
